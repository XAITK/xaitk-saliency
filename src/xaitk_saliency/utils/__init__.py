"""Package for xaitk-saliency utils needed for carrying out saliency generation"""
