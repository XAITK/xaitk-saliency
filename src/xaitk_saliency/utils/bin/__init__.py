"""Package for sal_on_coco_dets"""
