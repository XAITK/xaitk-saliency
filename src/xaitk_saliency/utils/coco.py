import logging
from collections.abc import Generator

import numpy as np
from PIL import Image  # type: ignore

try:
    import kwcoco  # type: ignore

    is_usable = True
except ModuleNotFoundError:
    is_usable = False

LOG = logging.getLogger(__name__)


if not is_usable:
    LOG.warning(f"{__name__} requires additional dependencies, please install 'xaitk-saliency[tools]'")
else:

    def parse_coco_dset(
        dets_dset: kwcoco.CocoDataset,
    ) -> Generator[tuple[np.ndarray, np.ndarray, np.ndarray], None, None]:
        """
        Generate reference image, bounding box, and class score matrices, for
        use with an implementation of `GenerateObjectDetectorBlackboxSaliency`,
        from a `kwcoco.CocoDataset` object.
        Matrices are generated for each image in the dataset that has
        annotations associated with it.

        The output bounding boxes matrices are each of shape [nDets x 4].
        The output score matrices are each of shape [nDets x nClasses].
        The value of nClasses is extrapolated from the minimum and maximum
        category id values present in the dataset.

        :param dets_dset: `kwcoco.CocoDataset` object to parse.

        :return: A generator of reference images, corresponding bounding boxes,
            and scores as would be input to an implementation of
            `GenerateObjectDetectorBlackboxSaliency`.
        """
        # filter out images with no detections
        gid_to_aids = {key: value for (key, value) in dets_dset.gid_to_aids.items() if len(value) > 0}

        cids = [cat["id"] for cat in dets_dset.cats.values()]
        min_cid = min(cids)
        num_classes = max(cids) - min_cid + 1

        for _img_i, (img_id, det_ids) in enumerate(gid_to_aids.items()):
            bboxes = np.empty((0, 4))
            scores = np.empty((0, num_classes))

            for det_id in det_ids:
                ann = dets_dset.anns[det_id]

                if "score" in ann:
                    score_array = np.zeros(num_classes)
                    score_array[ann["category_id"] - min_cid] = ann["score"]
                elif "prob" in ann:
                    score_array = ann["prob"]
                # annotation has no scores
                else:
                    # do not include any non-prediction annotations
                    continue  # pragma: no cover

                scores = np.vstack((scores, score_array))

                x, y, w, h = ann["bbox"]
                bbox = [x, y, x + w, y + h]
                bboxes = np.vstack((bboxes, bbox))

            img_file = dets_dset.get_image_fpath(img_id)
            ref_img = np.asarray(Image.open(img_file))

            yield ref_img, bboxes, scores
