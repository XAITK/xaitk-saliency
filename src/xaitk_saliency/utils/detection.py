from typing import Optional

import numpy as np


def format_detection(
    bbox_mat: np.ndarray,
    classification_mat: np.ndarray,
    objectness: Optional[np.ndarray] = None,
) -> np.ndarray:
    """
    Combine detection and classification output, with optional objectness
    output, into the combined format required for
    :py:meth:`.GenerateDetectorProposalSaliency.generate` ``*_dets`` input
    parameters.

    We enforce some shape consistency so that we can create a valid output
    matrix. The input bounding box matrix should be of shape ``[nDets x 4]``,
    the classification matrix should be of shape ``[nDets x nClasses]``, and
    the objectness vector, if provided, should be of size ``nDets``.

    If an ``objectness`` score vector is not provided, we assume a vector of
    1's.

    The output of this function is a matrix that is of shape
    ``[nDets x (4+1+nClasses)]``. This is the result of horizontally stacking the
    input in bbox, objectness and classification order. The output matrix
    data-type will follow numpy's rules about safe-casting given the
    combination of input matrix types.

    In exceptions about shape mismatches, index 0 refers to the ``bbox_mat``
    input, index 1 refers to the ``objectness`` vector, and index 2 refers to
    the ``classification_mat``.

    :param bbox_mat: Matrix of bounding boxes. This matrix should have the
        shape ``[nDets x 4]``. The format of each row-vector is not important
        but generally expected to be ``[left, top, right, bottom]`` pixel
        coordinates. This matrix must be of a type that is float-castable.
    :param classification_mat: Matrix of classification scores from the
        detector or detection classifier. This should have the shape of
        ``[nDets x nClasses]``. This matrix must be of a type that is
        float-castable.
    :param objectness: Optional vector of objectness scores for input
        detections. This is optional as not all detection models output this
        aspect. When provided, this should be a vector of ints/floats of size
        ``nDets`` to match the other parameter shapes.

    :raises ValueError: When input matrix shapes are mismatched such that they
        cannot be horizontally stacked.

    :returns: Matrix combining bounding box, objectness and class confidences.
    """
    if objectness is None:
        # Using the smallest sized type (boolean) to represent the 1-value so
        # as to not blowup the output type to something larger than input.
        objectness = np.full((bbox_mat.shape[0], 1), fill_value=True)
    elif objectness.ndim == 1:
        objectness = objectness.reshape(objectness.shape[0], 1)
    return np.hstack([bbox_mat, objectness, classification_mat])
