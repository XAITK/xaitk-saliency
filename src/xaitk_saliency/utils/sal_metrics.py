"""Utility functions for computing image saliency metrics."""

from __future__ import annotations

from typing import Any

import numpy as np
from scipy.ndimage import zoom
from skimage.transform import resize


def compute_ssd(sal_map: np.ndarray[Any, Any], ref_sal_map: np.ndarray[Any, Any]) -> float:
    """
    Computes the Sum of Squared Differences (SSD) between two saliency maps.

    Args:
        sal_map (np.ndarray): Predicted saliency map of shape (height, width).
        ref_sal_map (np.ndarray): Ground truth/reference saliency map of shape (height, width).

    Returns:
        float: The SSD value between the two saliency maps. If the norm is zero, returns np.inf.
    """
    sum_sq_diff = np.sum(np.power(np.subtract(sal_map, ref_sal_map), 2))
    norm = np.sqrt(np.sum(np.power(sal_map, 2)) * np.sum(np.power(ref_sal_map, 2)))
    if not norm:
        return np.inf
    return sum_sq_diff / norm


def compute_xcorr(sal_map: np.ndarray[Any, Any], ref_sal_map: np.ndarray[Any, Any]) -> float:
    """
    Computes the Normalized Cross-Correlation (NCC) between two saliency maps.

    Args:
        sal_map (np.ndarray): Predicted saliency map of shape (height, width).
        ref_sal_map (np.ndarray): Ground truth/reference saliency map of shape (height, width).

    Returns:
        float: The NCC value between the two saliency maps, ranging from [-1, 1].
    """

    def _normalize(s: np.ndarray[Any, Any]) -> tuple[np.ndarray[Any, Any], bool]:
        """Normalizes the saliency map by subtracting the mean and dividing by the standard deviation."""
        s -= s.mean()
        std = s.std()

        if std:
            s /= std

        return s, std == 0

    s1, c1 = _normalize(sal_map.copy())
    s2, c2 = _normalize(ref_sal_map.copy())

    if c1 and not c2:
        return 0.0
    return np.corrcoef(s1.flatten(), s2.flatten())[0, 1]


def compute_iou_coverage(
    saliency_features: np.ndarray[Any, Any],
    ground_truth_features: np.ndarray[Any, Any],
) -> float:
    """Returns the Shared Interest IoU Coverage metric for a single sample.

    IoU Coverage is the number of features in both the ground truth and saliency sets
    divided by the number of features in at least one of the ground truth and saliency
    sets. If one of the inputs is larger in spatial dimensions than the other, the larger
    input is scaled down to the size of the smaller input.

    Args:
        ground_truth_features (np.ndarray): Binary image mask for the GT class/object
            of shape (height, width).
        saliency_features (np.ndarray): Binary saliency map for the predicted class/object
            of shape (height, width).

    Returns:
        float: The computed IoU coverage metric.
    """

    # Determine the target dimensions (smallest among the two)
    target_h = min(ground_truth_features.shape[0], saliency_features.shape[0])
    target_w = min(ground_truth_features.shape[1], saliency_features.shape[1])
    target_shape = (target_h, target_w)

    def _scale_feature_to_target(
        feature: np.ndarray[Any, Any],
        target_shape: tuple[int, int],
    ) -> np.ndarray[Any, Any]:
        """Scales a single 2D feature map to the specified target shape using nearest-neighbor interpolation.

        Args:
            feature (np.ndarray): A 2D array of shape (height, width).
            target_shape (tuple): Target shape as (target_height, target_width).

        Returns:
            np.ndarray: Scaled feature of shape (target_height, target_width).
        """
        # scipy.zoom expects size as (width, height)
        x_zoom_factor = target_shape[1] / feature.shape[1]
        y_zoom_factor = target_shape[0] / feature.shape[0]
        return zoom(feature.astype(np.uint8), (x_zoom_factor, y_zoom_factor), order=1, grid_mode=True, mode="nearest")

    # Scale down if needed.
    if ground_truth_features.shape != target_shape:
        ground_truth_features = _scale_feature_to_target(
            ground_truth_features,
            target_shape,
        )
    if saliency_features.shape != target_shape:
        saliency_features = _scale_feature_to_target(saliency_features, target_shape)

    # Ensure the arrays are binary
    ground_truth_features = ground_truth_features.astype(bool)
    saliency_features = saliency_features.astype(bool)
    # Compute intersection and union
    intersection = np.sum(np.logical_and(ground_truth_features, saliency_features))
    union = np.sum(np.logical_or(ground_truth_features, saliency_features))
    return intersection / np.maximum(union, 1e-10)


def _downsample_to_target(
    arr: np.ndarray[Any, Any],
    target_shape: tuple[int, int],
    is_binary: bool = False,
) -> np.ndarray[Any, Any]:
    """Downsamples the input 2D array to the target spatial shape.

    Args:
        arr (np.ndarray): Input 2D array (height, width).
        target_shape (tuple[int, int]): Tuple consisting of target_height and target_width.
        is_binary (bool): If True, use nearest-neighbor (order=0) without anti-aliasing.

    Returns:
        np.ndarray: The resized array.
    """
    current_shape = arr.shape  # (height, width)
    if current_shape == target_shape:
        return arr
    order = 0 if is_binary else 1
    anti_aliasing = not is_binary
    # Resize the image. Note: preserve_range keeps the original value scale.
    return resize(
        arr,
        (target_shape[0], target_shape[1]),
        preserve_range=True,
        anti_aliasing=anti_aliasing,
        order=order,
    )


def compute_saliency_coverage(
    saliency_features: np.ndarray[Any, Any],
    ground_truth_features: np.ndarray[Any, Any],
) -> np.float64:
    """Returns the Shared Interest Saliency Coverage metric.

    The metric is computed as the saliency in the ground truth region divided by
    the total saliency. If the spatial dimensions of saliency_features and
    ground_truth_features differ, the larger is downsampled to match the smaller.

    Args:
        ground_truth_features (np.ndarray): Binary image mask for the GT class/object
            of shape (height, width).
        saliency_features (np.ndarray): Binary/Continuous saliency map for the predicted class/object
            of shape (height, width).

    Returns:
        numpy.float64: The computed saliency coverage metric.
    """
    # Determine the target spatial shape (the smaller dimensions between the two)
    target_h = min(saliency_features.shape[0], ground_truth_features.shape[0])
    target_w = min(saliency_features.shape[1], ground_truth_features.shape[1])
    target_shape = (target_h, target_w)

    # Downsample the larger array(s) if needed.
    # Assume ground_truth is binary and saliency_features may be continuous.
    if saliency_features.shape != target_shape:
        saliency_features = _downsample_to_target(
            saliency_features,
            target_shape,
            is_binary=False,
        )
    if ground_truth_features.shape != target_shape:
        ground_truth_features = _downsample_to_target(
            ground_truth_features,
            target_shape,
            is_binary=True,
        )

    # Compute the intersection and total saliency
    intersection = np.sum(ground_truth_features * saliency_features)
    explanation_saliency = np.sum(saliency_features)
    return intersection / max(explanation_saliency, 1e-10)


def compute_ground_truth_coverage(
    saliency_features: np.ndarray[Any, Any],
    ground_truth_features: np.ndarray[Any, Any],
) -> np.float64:
    """Returns the Shared Interest Ground Truth Coverage metric.

    The metric is computed as the number of features that occur in both the ground truth
    and saliency feature sets divided by the number of ground truth features.
    If the spatial dimensions differ, the larger array is downsampled to the size of the smaller.

    Args:
        ground_truth_features (np.ndarray): Binary image mask for the GT class/object
            of shape (height, width).
        saliency_features (np.ndarray): Binary saliency map for the predicted class/object
            of shape (height, width).

    Returns:
        numpy.float64: The computed ground truth coverage metric.
    """
    target_h = min(saliency_features.shape[0], ground_truth_features.shape[0])
    target_w = min(saliency_features.shape[1], ground_truth_features.shape[1])
    target_shape = (target_h, target_w)

    # Both inputs are binary; use nearest neighbor resizing.
    if saliency_features.shape[1:] != target_shape:
        saliency_features = _downsample_to_target(
            saliency_features,
            target_shape,
            is_binary=False,
        )
    if ground_truth_features.shape[1:] != target_shape:
        ground_truth_features = _downsample_to_target(
            ground_truth_features,
            target_shape,
            is_binary=True,
        )

    # Compute the intersection and the total ground truth features.
    # Here we sum over all spatial dimensions.
    intersection = np.sum(ground_truth_features * saliency_features)
    ground_truth_saliency = np.sum(ground_truth_features)
    return intersection / max(ground_truth_saliency, 1e-10)
