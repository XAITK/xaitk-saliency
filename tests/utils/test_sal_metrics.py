from __future__ import annotations

import copy
from typing import Callable

import numpy as np
import pytest
from syrupy.assertion import SnapshotAssertion

from tests.test_utils import CustomFloatSnapshotExtension
from xaitk_saliency.utils.sal_metrics import (
    compute_ground_truth_coverage,
    compute_iou_coverage,
    compute_saliency_coverage,
    compute_ssd,
    compute_xcorr,
)

# Define two different random number generator seeds
rng_1 = np.random.default_rng(123)
rng_2 = np.random.default_rng(456)

# Define two different Gumbel distribution-based saliency maps
# Note: Gumbel distribution is used to model extreme-values which provide
# a fair approximation for high/low salient feature maps
predicted_saliency_map = rng_1.gumbel(loc=0, scale=0.1, size=(256, 256))
reference_saliency_map = rng_2.gumbel(loc=0, scale=0.1, size=(256, 256))

# Define binary image mask and binary/continuous saliency maps
gt_binary_mask = rng_2.integers(0, 2, size=(256, 256))
saliency_binary_map = rng_1.integers(0, 2, size=(256, 256))
saliency_continous_map = predicted_saliency_map

# Define downsized image mask and saliency maps to test the feature-scaling use-case
gt_downsized_binary_mask = rng_2.integers(0, 2, size=(128, 128))
saliency_downsized_binary_mask = rng_1.integers(0, 2, size=(128, 128))
saliency_downsized_continous_map = rng_1.gumbel(loc=0, scale=0.1, size=(128, 128))


def saliency_metric_assertions(
    computation: Callable,
    sal_map: np.ndarray,
    ref_sal_map: np.ndarray,
) -> None:
    """Test that the inputs are not modified while computing an image metric.

    :param computation: Metric computation function to test.
    :param sal_map: Input saliency map.
    :param ref_sal_map: Reference saliency map for comparison.
    """
    original_sal_map = copy.deepcopy(sal_map)
    original_ref_sal_map = copy.deepcopy(ref_sal_map)
    original_metric_value = computation(original_sal_map, original_ref_sal_map)

    assert np.array_equal(original_sal_map, sal_map), "sal_map modified, data changed"
    assert np.array_equal(original_ref_sal_map, ref_sal_map), "ref_sal_map modified, data changed"

    metric_value = computation(sal_map, ref_sal_map)

    assert original_metric_value == metric_value, "metric_value modified, data changed"


@pytest.fixture
def snapshot_custom(snapshot: SnapshotAssertion) -> SnapshotAssertion:
    return snapshot.use_extension(CustomFloatSnapshotExtension)


class TestComputeSSD:
    """This class contains the unit tests for the functionality of the SSD metric util function."""

    @pytest.mark.parametrize(
        ("sal_map", "ref_sal_map"),
        [(predicted_saliency_map, reference_saliency_map)],
        ids=["Single random input saliency and reference saliency maps"],
    )
    def test_consistency(
        self,
        sal_map: np.ndarray,
        ref_sal_map: np.ndarray,
    ) -> None:
        """Test SSD metric util function with various random saliency map inputs."""

        saliency_metric_assertions(
            computation=compute_ssd,
            sal_map=sal_map,
            ref_sal_map=ref_sal_map,
        )

    @pytest.mark.parametrize(
        ("sal_map", "ref_sal_map"),
        [(predicted_saliency_map, reference_saliency_map)],
        ids=["Single random input saliency and reference saliency maps"],
    )
    def test_regression(
        self,
        sal_map: np.ndarray,
        ref_sal_map: np.ndarray,
        snapshot_custom: SnapshotAssertion,
    ) -> None:
        """Regression testing results to detect API changes."""
        ssd_metric_value = compute_ssd(sal_map=sal_map, ref_sal_map=ref_sal_map)
        snapshot_custom.assert_match(ssd_metric_value)


class TestComputeXCorr:
    """This class contains the unit tests for the functionality of the XCorr metric util function."""

    @pytest.mark.parametrize(
        ("sal_map", "ref_sal_map"),
        [(predicted_saliency_map, reference_saliency_map)],
        ids=["Single random input saliency and reference saliency maps"],
    )
    def test_consistency(
        self,
        sal_map: np.ndarray,
        ref_sal_map: np.ndarray,
    ) -> None:
        """Test XCorr metric util function with various random saliency map inputs."""

        saliency_metric_assertions(
            computation=compute_xcorr,
            sal_map=sal_map,
            ref_sal_map=ref_sal_map,
        )

    @pytest.mark.parametrize(
        ("sal_map", "ref_sal_map"),
        [(predicted_saliency_map, reference_saliency_map)],
        ids=["Single random input saliency and reference saliency maps"],
    )
    def test_regression(
        self,
        sal_map: np.ndarray,
        ref_sal_map: np.ndarray,
        snapshot_custom: SnapshotAssertion,
    ) -> None:
        """Regression testing results to detect API changes."""
        xcorr_metric_value = compute_xcorr(sal_map=sal_map, ref_sal_map=ref_sal_map)
        snapshot_custom.assert_match(xcorr_metric_value)


class TestComputeGroundTruthCoverage:
    """This class contains the unit tests for the functionality of the Ground Truth coverage metric util function."""

    @pytest.mark.parametrize(
        ("saliency_features", "ground_truth_features"),
        [
            (saliency_binary_map, gt_binary_mask),
            (saliency_downsized_binary_mask, gt_binary_mask),
            (saliency_binary_map, gt_downsized_binary_mask),
        ],
        ids=[
            "Single random binary saliency map and image mask",
            "Single random downsized binary saliency map and image mask",
            "Single random binary saliency map and downsized image mask",
        ],
    )
    def test_consistency(
        self,
        saliency_features: np.ndarray,
        ground_truth_features: np.ndarray,
    ) -> None:
        """Test Ground Truth Coverage metric util function with various random saliency map inputs."""

        saliency_metric_assertions(
            computation=compute_ground_truth_coverage,
            sal_map=saliency_features,
            ref_sal_map=ground_truth_features,
        )

    @pytest.mark.parametrize(
        ("saliency_features", "ground_truth_features"),
        [
            (saliency_binary_map, gt_binary_mask),
            (saliency_downsized_binary_mask, gt_binary_mask),
            (saliency_binary_map, gt_downsized_binary_mask),
        ],
        ids=[
            "Single random binary saliency map and image mask",
            "Single random downsized binary saliency map and image mask",
            "Single random binary saliency map and downsized image mask",
        ],
    )
    def test_regression(
        self,
        saliency_features: np.ndarray,
        ground_truth_features: np.ndarray,
        snapshot_custom: SnapshotAssertion,
    ) -> None:
        """Regression testing results to detect API changes."""
        gt_coverage_metric_value = compute_ground_truth_coverage(
            saliency_features=saliency_features,
            ground_truth_features=ground_truth_features,
        )
        snapshot_custom.assert_match(gt_coverage_metric_value)


class TestComputeSaliencyCoverage:
    """This class contains the unit tests for the functionality of the Saliency coverage metric util function."""

    @pytest.mark.parametrize(
        ("saliency_features", "ground_truth_features"),
        [
            (saliency_binary_map, gt_binary_mask),
            (saliency_downsized_binary_mask, gt_binary_mask),
            (saliency_binary_map, gt_downsized_binary_mask),
            (saliency_continous_map, gt_binary_mask),
            (saliency_downsized_continous_map, gt_binary_mask),
            (saliency_continous_map, gt_downsized_binary_mask),
        ],
        ids=[
            "Single random binary saliency map and image mask",
            "Single random downsized binary saliency map and image mask",
            "Single random binary saliency map and downsized image mask",
            "Single random continuous saliency map and binary image mask",
            "Single random downsized continuous saliency map and binary image mask",
            "Single random continuous saliency map and downsized binary image mask",
        ],
    )
    def test_consistency(
        self,
        saliency_features: np.ndarray,
        ground_truth_features: np.ndarray,
    ) -> None:
        """Test Saliency coverage metric util function with various random saliency map inputs."""

        saliency_metric_assertions(
            computation=compute_saliency_coverage,
            sal_map=saliency_features,
            ref_sal_map=ground_truth_features,
        )

    @pytest.mark.parametrize(
        ("saliency_features", "ground_truth_features"),
        [
            (saliency_binary_map, gt_binary_mask),
            (saliency_downsized_binary_mask, gt_binary_mask),
            (saliency_binary_map, gt_downsized_binary_mask),
            (saliency_continous_map, gt_binary_mask),
            (saliency_downsized_continous_map, gt_binary_mask),
            (saliency_continous_map, gt_downsized_binary_mask),
        ],
        ids=[
            "Single random binary saliency map and image mask",
            "Single random downsized binary saliency map and image mask",
            "Single random binary saliency map and downsized image mask",
            "Single random continuous saliency map and binary image mask",
            "Single random downsized continuous saliency map and binary image mask",
            "Single random continuous saliency map and downsized binary image mask",
        ],
    )
    def test_regression(
        self,
        saliency_features: np.ndarray,
        ground_truth_features: np.ndarray,
        snapshot_custom: SnapshotAssertion,
    ) -> None:
        """Regression testing results to detect API changes."""
        saliency_coverage_metric_value = compute_saliency_coverage(
            saliency_features=saliency_features,
            ground_truth_features=ground_truth_features,
        )
        snapshot_custom.assert_match(saliency_coverage_metric_value)


class TestComputeIoUCoverage:
    """This class contains the unit tests for the functionality of the IoU coverage metric util function."""

    @pytest.mark.parametrize(
        ("saliency_features", "ground_truth_features"),
        [
            (saliency_binary_map, gt_binary_mask),
            (saliency_downsized_binary_mask, gt_binary_mask),
            (saliency_binary_map, gt_downsized_binary_mask),
        ],
        ids=[
            "Single random binary saliency map and image mask",
            "Single random downsized binary saliency map and image mask",
            "Single random binary saliency map and downsized image mask",
        ],
    )
    def test_consistency(
        self,
        saliency_features: np.ndarray,
        ground_truth_features: np.ndarray,
    ) -> None:
        """Test IoU coverage metric util function with various random saliency map inputs."""

        saliency_metric_assertions(
            computation=compute_iou_coverage,
            sal_map=saliency_features,
            ref_sal_map=ground_truth_features,
        )

    @pytest.mark.parametrize(
        ("saliency_features", "ground_truth_features"),
        [
            (saliency_binary_map, gt_binary_mask),
            (saliency_downsized_binary_mask, gt_binary_mask),
            (saliency_binary_map, gt_downsized_binary_mask),
        ],
        ids=[
            "Single random binary saliency map and image mask",
            "Single random downsized binary saliency map and image mask",
            "Single random binary saliency map and downsized image mask",
        ],
    )
    def test_regression(
        self,
        saliency_features: np.ndarray,
        ground_truth_features: np.ndarray,
        snapshot_custom: SnapshotAssertion,
    ) -> None:
        """Regression testing results to detect API changes."""
        iou_coverage_metric_value = compute_iou_coverage(
            saliency_features=saliency_features,
            ground_truth_features=ground_truth_features,
        )
        snapshot_custom.assert_match(iou_coverage_metric_value)
