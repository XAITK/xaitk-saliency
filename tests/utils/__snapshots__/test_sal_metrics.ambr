# serializer version: 1
# name: TestComputeGroundTruthCoverage.test_regression[Single random binary saliency map and downsized image mask]
  0.5036385690673137
# ---
# name: TestComputeGroundTruthCoverage.test_regression[Single random binary saliency map and image mask]
  0.5038462717214748
# ---
# name: TestComputeGroundTruthCoverage.test_regression[Single random downsized binary saliency map and image mask]
  0.4919157275845174
# ---
# name: TestComputeIoUCoverage.test_regression[Single random binary saliency map and downsized image mask]
  0.41187518058364636
# ---
# name: TestComputeIoUCoverage.test_regression[Single random binary saliency map and image mask]
  0.33478597320083087
# ---
# name: TestComputeIoUCoverage.test_regression[Single random downsized binary saliency map and image mask]
  0.401999855082965
# ---
# name: TestComputeSSD.test_regression[Single random input saliency and reference saliency maps]
  1.6505363055162763
# ---
# name: TestComputeSaliencyCoverage.test_regression[Single random binary saliency map and downsized image mask]
  0.5058850933481778
# ---
# name: TestComputeSaliencyCoverage.test_regression[Single random binary saliency map and image mask]
  0.4994379803748823
# ---
# name: TestComputeSaliencyCoverage.test_regression[Single random continuous saliency map and binary image mask]
  0.49620540669131535
# ---
# name: TestComputeSaliencyCoverage.test_regression[Single random continuous saliency map and downsized binary image mask]
  0.5108586717850373
# ---
# name: TestComputeSaliencyCoverage.test_regression[Single random downsized binary saliency map and image mask]
  0.4934267108981447
# ---
# name: TestComputeSaliencyCoverage.test_regression[Single random downsized continuous saliency map and binary image mask]
  0.5114136755802363
# ---
# name: TestComputeXCorr.test_regression[Single random input saliency and reference saliency maps]
  0.005532439020454301
# ---
